# Building a Machine Learning Model to predict Breast Cancer Data

 Nowadays breast cancer is a common disease in whole worldwide. New cancer cases and cancer-related deaths according to global statistics showing that it a significant public health problem in today’s society. But the early diagnosis can improve the prognosis and chance of survival significantly. Furthermore, accurate classification of benign tumors can prevent patients undergoing unnecessary treatments. Thus, the correct diagnosis and classification of patients into malignant/ benign groups is the subject of much research. Because of its unique advantages in critical features detection from complex datasets, machine learning is widely recognized as the methodology to choice pattern classification and forecast modelling.


For dataset, [Click here for the dataset](https://gist.github.com/KhanradCoder/35a6beea49e5b9ba62797e595a9626c0)

 * Data Exploration and Preparation

In [1]:
import pandas as pd

dataset = pd.read_csv('cancer.csv')
dataset.head(10)

diagnosis(1=m, 0=b)  radius_mean  texture_mean  perimeter_mean  area_mean  \
0                    1        17.99         10.38          122.80     1001.0   
1                    1        20.57         17.77          132.90     1326.0   
2                    1        19.69         21.25          130.00     1203.0   
3                    1        11.42         20.38           77.58      386.1   
4                    1        20.29         14.34          135.10     1297.0   
5                    1        12.45         15.70           82.57      477.1   
6                    1        18.25         19.98          119.60     1040.0   
7                    1        13.71         20.83           90.20      577.9   
8                    1        13.00         21.82           87.50      519.8   
9                    1        12.46         24.04           83.97      475.9   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760         0.30010              0.14710   
1          0.08474           0.07864         0.08690              0.07017   
2          0.10960           0.15990         0.19740              0.12790   
3          0.14250           0.28390         0.24140              0.10520   
4          0.10030           0.13280         0.19800              0.10430   
5          0.12780           0.17000         0.15780              0.08089   
6          0.09463           0.10900         0.11270              0.07400   
7          0.11890           0.16450         0.09366              0.05985   
8          0.12730           0.19320         0.18590              0.09353   
9          0.11860           0.23960         0.22730              0.08543   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   
5         0.2087  ...         15.47          23.75           103.40   
6         0.1794  ...         22.88          27.66           153.20   
7         0.2196  ...         17.06          28.14           110.60   
8         0.2350  ...         15.49          30.73           106.20   
9         0.2030  ...         15.09          40.68            97.65   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   
5       741.6            0.1791             0.5249           0.5355   
6      1606.0            0.1442             0.2576           0.3784   
7       897.0            0.1654             0.3682           0.2678   
8       739.3            0.1703             0.5401           0.5390   
9       711.4            0.1853             1.0580           1.1050   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  
5                0.1741          0.3985                  0.12440  
6                0.1932          0.3063                  0.08368  
7                0.1556          0.3196                  0.11510  
8                0.2060          0.4378                  0.10720  
9                0.2210          0.4366 

In [2]:
# identify features and level
x = dataset.drop(columns=["diagnosis(1=m, 0=b)"])
y = dataset["diagnosis(1=m, 0=b)"]

* Split data into train and test set.

In [3]:
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(x, y, test_size=0.25, random_state = 0)

In [4]:
print(X_train.shape)
print(X_test.shape)

(426, 30)
(143, 30)


In [5]:
#Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

* Model Selection

In [6]:
#Using Logistic Regression Algorithm to the Training Set
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, Y_train)

LogisticRegression(random_state=0)

In [7]:
#Using KNeighborsClassifier Method of neighbors class to use Nearest Neighbor algorithm
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifier.fit(X_train, Y_train)

KNeighborsClassifier()

In [8]:
#Using SVC method of svm class to use Support Vector Machine Algorithm
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state = 0)
classifier.fit(X_train, Y_train)

SVC(kernel='linear', random_state=0)

In [9]:
#Using SVC method of svm class to use Kernel SVM Algorithm
from sklearn.svm import SVC
classifier = SVC(kernel = 'rbf', random_state = 0)
classifier.fit(X_train, Y_train)

SVC(random_state=0)

In [10]:
#Using GaussianNB method of naïve_bayes class to use Naïve Bayes Algorithm
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, Y_train)

GaussianNB()

In [11]:
#Using DecisionTreeClassifier of tree class to use Decision Tree Algorithm

from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train, Y_train)

DecisionTreeClassifier(criterion='entropy', random_state=0)

In [12]:
#Using RandomForestClassifier method of ensemble class to use Random Forest Classification algorithm

from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
classifier.fit(X_train, Y_train)

RandomForestClassifier(criterion='entropy', n_estimators=10, random_state=0)

* Evaluation

In [13]:
Y_pred = classifier.predict(X_test)
#Creating the confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, Y_pred)
print(cm)
accuracy=(cm[0,0]+cm[1,1])/((cm[0,0]+cm[1,1])+(cm[1,0]+cm[0,1]))*100
print("Accuracy for cancer preiction: {:.1f}%".format(accuracy))


[[89  1]
 [ 1 52]]
Accuracy for cancer preiction: 98.6%


In [14]:
import pickle
 
pickle.dump(classifier, open('modelRF.pkl', 'wb'))

print("pkl file saved")

pkl file saved


In [15]:
# Loading model to compare the results
model = pickle.load(open('modelRF.pkl','rb'))
print(model.predict([[.2, .9, .6,.2,2,.3,.3,.3,.6,.5,
                      .4,.1,.4,.1,.4,.1,.7,.7,.8,.5,
                      .5, .5,.4,.4,.8,.8,.9,.6,.5,.4]]))

[1]


## After applying the different classification models, we have got below accuracies with different models:

1. Logistic Regression — 95.8%

2. Nearest Neighbor — 95.1%

3. Support Vector Machines — 97.2%

4. Kernel SVM — 96.5%

5. Naive Bayes — 91.6%

6. Decision Tree Algorithm — 95.8%

7. Random Forest Classification — 98.6%